beluga/include/beluga/algorithm/effective_sample_size.hpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ALGORITHM_EFFECTIVE_SAMPLE_SIZE_HPP
+#define BELUGA_ALGORITHM_EFFECTIVE_SAMPLE_SIZE_HPP
+
+#include <beluga/type_traits/particle_traits.hpp>
+#include <beluga/views/particles.hpp>
+
+#include <range/v3/numeric/accumulate.hpp>
+
+/**
+ * \file
+ * \brief Implementation of an algorithm to calculate the effective sample size (ESS).
+ */
+
+namespace beluga {
+
+/// Calculate the ESS of a given a range of weights.
+/**
+ * The effective sample size (ESS) is a metric that measures how much information content is loss due to the
+ * correlation in the sequence. In other words, although our sequence may have a length (sample size) of N,
+ * our effective sample size will be smaller due to the correlation and redundancy between the samples.
+ *
+ * When compared with the sample size, it can be used as a measure of weight degeneracy.
```

Comment:
@nahueespinosa meta: code looks good but the explanation seems off. IIUC, ESS is an estimate of the particle filter's efficiency as an estimator. Higher weight (or likelihood or unnormalized probability) dispersion means lower efficiency, thus lower ESS. Correlation and redundancy among samples have nothing to do with that. Section 4.1 of [Liu et. al.](https://statweb.rutgers.edu/rongchen/papers/KongLiuWong.pdf) (the seminal paper on ESS) is a bit cryptic but conveys just that IIUC.

Reply:
Yeah, I guess some definitions in the interwebs (?) confused me. Do you have a suggestion to fix the description (or should I just remove it)?

Reply:
Hmm, how about:
```suggestion
 * The effective sample size (ESS) is a figure of merit for importance sampling methods' output. It characterizes
 * how well the target posterior distribution is approximated (as it is proportional to the efficiency of  distribution
 * parameter estimators <insert liu et. al. reference, section 4.1>). It can be interpreted as the number of samples
 * effectively approximating the distribution, and thus comparing with the total number of samples makes for a good
 * mechanism to detect and react to performance degradation e.g. triggering a resample when the ESS falls below a 
 * fraction of the total number of samples.
```

The other references are also useful.

Reply:
Done! https://github.com/Ekumen-OS/beluga/pull/290/commits/8a7e8e0163540e41761e825028ab4040070febe9.